contracts/staking/StakeHolder.sol

```diff
@@ -0,0 +1,266 @@
+// Copyright (c) Immutable Pty Ltd 2018 - 2024
+// SPDX-License-Identifier: Apache 2
+pragma solidity ^0.8.19;
+
+import {UUPSUpgradeable} from "openzeppelin-contracts-upgradeable-4.9.3/proxy/utils/UUPSUpgradeable.sol";
+import {AccessControlEnumerableUpgradeable} from "openzeppelin-contracts-upgradeable-4.9.3/access/AccessControlEnumerableUpgradeable.sol";
+
+/// @notice Struct to combine an account and an amount.
+struct AccountAmount {
+    address account;
+    uint256 amount;
+}
+
+/**
+ * @title StakeHolder: allows anyone to stake any amount of native IMX and to then remove all or part of that stake.
+ * @dev The StakeHolder contract is designed to be upgradeable.
+ */
+contract StakeHolder is AccessControlEnumerableUpgradeable, UUPSUpgradeable {
+    /// @notice Error: Attempting to upgrade contract storage to version 0.
+    error CanNotUpgradeToLowerOrSameVersion(uint256 _storageVersion);
+
+    /// @notice Error: Attempting to renounce the last role admin / default admin.
+    error MustHaveOneRoleAdmin();
+
+    /// @notice Error: Attempting to stake with zero value.
+    error MustStakeMoreThanZero();
+
+    /// @notice Error: Attempting to distribute zero value.
+    error MustDistributeMoreThanZero();
+
+    /// @notice Error: Attempting to unstake amount greater than the balance.
+    error UnstakeAmountExceedsBalance(uint256 _amountToUnstake, uint256 _currentStake);
+
+    /// @notice Error: The sum of all amounts to distribute did not equal msg.value of the distribute transaction.
+    error DistributionAmountsDoNotMatchTotal(uint256 _msgValue, uint256 _calculatedTotalDistribution);
+
+    /// @notice Error: Distributions can only be made to accounts that have staked.
+    error AttemptToDistributeToNewAccount(address _account, uint256 _amount);
+
+    /// @notice Event when an amount has been staked or when an amount is distributed to an account.
+    event StakeAdded(address _staker, uint256 _amountAdded, uint256 _newBalance);
+
+    /// @notice Event when an amount has been unstaked.
+    event StakeRemoved(address _staker, uint256 _amountRemoved, uint256 _newBalance);
+
+    /// @notice Event summarising a distribution. There will also be one StakeAdded event for each recipient.
+    event Distributed(address _distributor, uint256 _totalDistribution, uint256 _numRecipients);
+
+    /// @notice Only UPGRADE_ROLE can upgrade the contract
+    bytes32 public constant UPGRADE_ROLE = bytes32("UPGRADE_ROLE");
+
+    /// @notice Version 0 version number
+    uint256 private constant _VERSION0 = 0;
+
+    /// @notice Holds staking information for a single staker.
+    struct StakeInfo {
+        /// @notice Amount of stake.
+        uint256 stake;
+        /// @notice True if this account has ever staked.
+        bool hasStaked;
+    }
+
+    /// @notice The amount of value owned by each staker
+    // solhint-disable-next-line private-vars-leading-underscore
```

Comment:
Turning this off globally is probably what we should do. However, let's not do it in this PR.
Using _ for variables that are supposed to be private is so that it is obvious if someone mistakenly makes a variable public. 
For me, it is a stylistic thing which I don't like. When I see _var I automatically assume it is a function parameter. Most Solidity written pre-circa 2020 used that style. I assume many other people who have used Solidity for a long time also have that reaction.